Ruby - Project My Enumerables

expjazz opened the pull request:
#### What does this project do?
It presents various methods for iterating through arrays.

### Description of Task to be completed?
 - Added the main.rb file.
 - Created all the methods.
 - Ran validations on Rubocop.
 - Rubocop was showing an error of Perceived/Complexity on the inject method that I had to ignore since I needed the method as it is to work.

#### How should this be manually tested?
- Clone this branch.
- open the main.rb file on your browser

#### Screenshot

![screenshot](https://user-images.githubusercontent.com/61479648/79880906-f8411700-83c6-11ea-86da-641ac045dd52.png)

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+  def my_each
+    i = 0
+    if block_given?
+      while i <= length - 1
```
Well I think you can make the condition like 
```ruby
while i < length 
```
It looks a little bit shorter. That's it.

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+
+  def my_select
+    arr = []
+    return to_enum if block_given? == false
```
You can use the `unless` instead of `if`. Then the code will be simpler than it is now.
```ruby
return to_enum unless block_given?
```

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+  end
+
+  def my_all?
+    return to_enum unless block_given?
```
As far I know all? does not return an enumerator if the block is got given. It adds an implicit block(well you can say checks none of the items are nil/false).

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+
+    my_each { |x| return false if yield(x) == false }
+    true
+  end
```
`my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) 
- when no block or argument is given, returns true when none of the collection members are false or nil
```ruby
[1, 2, 3].my_all? #should return true
[1, 2, nil].my_all? #should return false
[1, false, nil].my_all? #should return false
```
- when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
[1, 2, 3].my_all?(Integer) #should return true
[1, 2, nil].my_all?(Integer) #should return false
```
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
%w[dog door rod blade].my_all?(/d/)   #should return true.
%w[dog door rod blade].my_all?(/t/)   #should return false.
```
- when a pattern other than Regex or a Class is given, returns true if all of the collection matches the pattern
```ruby
[3, 3, 3].my_all?(3) #should return true
[3, 3, 1].my_all?(3) #should return false
```

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+  end
+
+  def my_any?(&prc)
+    return to_enum unless block_given?
```
It does not return an enumerator. If the block is not given, adds an implicit block(return true if at least one of the collection members is not false or nil.)

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+
+    my_each { |x| return true if prc.call(x) }
+    false
+  end
```
`my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
[1, 2, 3].my_any? #should return true
[1, 2, nil].my_any? #should return true
[false, false, nil].my_any? #should return false
```
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
[1, 2, 3].my_any?(Integer) #should return true
[1, 'demo', false].my_any?(Integer) #should return true
['demo', false, nil].my_any?(Integer) #should return false
```
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_any?(/t/)    #should return false.
%w[dog door rod blade].my_any?(/d/)   #should return true.
```
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
[1, 2, 3].my_any?(3)   #should return true
[1, 2, 2].my_any?(3)   #should return false
```

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+  end
+
+  def my_none?
+    return to_enum if block_given? == false
```
It does not return an enumerator. If the block is not given, none? will return true only if none of the collection members is true.

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
 
+    my_each { |x| return false if yield(x) == true }
+    true
+  end
```
`my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when no block or argument is given returns true only if none of the collection members is true
```ruby
[false, nil, false].my_none? #should return true
[1, 'demo', 2.2].my_none? #should return false
```
- when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
[1, 2, 3].my_none?(String)  #should return true
['1', '2', '3'].my_none?(String)  #should return false
```
- when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_none?(/z/)   #should return true
%w[dog door rod blade].my_none?(/d/)   #should return false
```
- when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
[1, 2, 4].my_none?(3) #should return true
[1, 2, 3].my_none?(3) #should return false
```

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+  end
+
+  def my_count(&prc)
+    return to_enum unless block_given?
```
If a block is given, it counts the number of elements yielding a true value. Not an enumerator.

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+    count = 0
+    my_each { |x| count += 1 if prc.call(x) == true }
+    count
+  end
```
`my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- returns the number of items in enum through enumeration
```ruby
[1, 2, 3].my_count  #should return 3
```
- counts the number of items in enum that are equal to item if an argument is given
```ruby
[1, 1, 2 ,3].my_count(1)  #should return 2
```

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+      my_each { |x| arr << prc.call(x) }
+    end
+    arr
+  end
```
`my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- returns an Enumerator if no block is given

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,91 @@
+      end
+    end
+    self[0]
+  end
```
`my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
```
- when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
[1, 2, 3].my_inject(:+)  #should return 6
[1, 2, 3].my_inject(1, :+)  #should return 7
```

expjazz commented:
## Made changes required by TSE
- while loop updated
- if instead of unless on my_select
- corrected return of my_all? on all cases
- corrected return of my_any?
- corrected return of my_none?
- corrected return of my_count
- updated my_inject and my_map as requested
- added multiply_els method

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,140 @@
+    elsif arg.nil? == false
+      my_each { |x| return false if x != arg }
+    else
+      my_each { |x| return false if x.nil? == true }
```
When I tried with a different case found a failing edge
```ruby
[1, false, 'hi', []].my_all?   #should return false but your current formation returns true.
```
Cause It checks for nil. When we check false.nil? it returns false but you are expecting it should be true.
![image](https://user-images.githubusercontent.com/38625956/79918496-51a05900-844e-11ea-8f75-00bac6562b0c.png)
So to cover that scenario I think the following change can do so.
```ruby
 my_each { |x| return false if x.nil? == true }
```
should be
 ```ruby
 my_each { |x| return false unless x } 
```

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,140 @@
+    true
+  end
+
+  def my_any?(&prc)
+    if block_given?
+      my_each { |x| return true if prc.call(x) }
+    else
+      my_each { |x| return true if x.nil? == false }
+    end
+    false
+  end
```
I don't think that you have covered the cases added in the previous comment. Let me put them again->
`my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
[1, 2, 3].my_any? #should return true
[1, 2, nil].my_any? #should return true
[false, false, nil].my_any? #should return false
```
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
[1, 2, 3].my_any?(Integer) #should return true
[1, 'demo', false].my_any?(Integer) #should return true
['demo', false, nil].my_any?(Integer) #should return false
```
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_any?(/t/)    #should return false.
%w[dog door rod blade].my_any?(/d/)   #should return true.
```
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
[1, 2, 3].my_any?(3)   #should return true
[1, 2, 2].my_any?(3)   #should return false
```

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,140 @@
+    elsif arg.nil? == false
+      my_each { |x| return false if x == arg }
+    else
+      my_each { |x| return false if x.nil? == false }
```
when no block or argument is given returns true only if none of the collection members is true. Here you are also comparing it with nil. What if we have a false? I think
```ruby
my_each { |x| return false if x.nil? == false }
```
should be 
```ruby
my_each { |x| return false if x }
```

shshamim63 commented on main.rb:
```diff
@@ -1,17 +1,140 @@
+  end
+
+  def my_inject(arg = nil, arg_2 = nil)
+
+    ind = 0
+    if block_given? == false
+      if arg.is_a?(Symbol)
+        while ind < length - 1
+          self[0] = self[0] + self[ind + 1] if arg[0] == '+'
+          self[0] = self[0] - self[ind + 1] if arg[0] == '-'
+          self[0] = self[0] * self[ind + 1] if arg[0] == '*'
+          self[0] = self[0] / self[ind + 1] if arg[0] == '/'
+          ind += 1
+        end
+      elsif arg.is_a?(Integer)
+        while ind < length - 1
+          self[0] = self[0] + arg if arg_2[0] == '+' && ind.zero?
+          self[0] = self[0] + self[ind + 1] if arg_2[0] == '+'
+          self[0] = self[0] - arg if arg_2[0] == '-' && ind.zero?
+          self[0] = self[0] - self[ind + 1] if arg_2[0] == '-'
+          self[0] = self[0] * arg if arg_2[0] == '*' && ind.zero?
+          self[0] = self[0] * self[ind + 1] if arg_2[0] == '*'
+          self[0] = self[0] / arg if arg_2[0] == '/' && ind.zero?
+          self[0] = self[0] / self[ind + 1] if arg_2[0] == '/'
+          ind += 1
+        end
+      end
+    elsif block_given?
+      if arg.nil?
+        my_each_with_index do |_ele, ind|
+          self[0] = yield(self[0], self[ind + 1]) if ind < length - 1
         end
-        self
+      else
+        my_each_with_index do |_ele, ind|
+          self[0] = yield(num_1, self[0]) if ind.zero?
+          self[0] = yield(self[0], self[ind]) if ind.positive?
+        end
+      end
     end
```
Well, it is partially functioning. Let's dive into a little bit depth to make it fully functional.
- inject can accept a proc. And based on that proc it can generate the result.
```ruby
LOWEST_VALUE = 0
HIGHEST_VALUE = 9
ARRAY_SIZE = 10
array = Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) }
operation = proc { |sum, n| sum + n }
array.my_inject(&operation)   #this should also perform properly.
```
- inject can perform on range too.
```ruby
(1...50).my_inject(4) { |prod, n| prod * n } #should also perform properly
```

expjazz commented:
## Made all the changes that were required (including the ones I should have made after the first review)

- Shout out to @shshamim63 for the amazing review and for the tips! 

- Here is a screenshot covering all previous test cases that my methods were failing (hopefully they are working now)
![my_any?testcases](https://user-images.githubusercontent.com/61479648/79963559-f24b4480-845f-11ea-8d12-9f077d98e7eb.png)

Torres-ssf commented on main.rb:
```diff
@@ -1,17 +1,160 @@
+    false
+  end
+
+  def my_none?(arg = nil, &prc)
+    if block_given?
+      my_each { |x| return false if prc.call(x) == true }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return false if arg.match?(x.to_s) == true }
+    elsif arg.is_a?(Class)
+      my_each { |x| return false if x.is_a?(arg) == true }
```
- my_none?

- [ ] when no block or argument is given returns true only if none of the collection members is true. Check out [this documentation](https://apidock.com/ruby/Enumerable/none%3F)
`[false, nil].my_none?` should returns true
![Screenshot from 2020-04-22 08-52-45](https://user-images.githubusercontent.com/30977845/79978681-ac4dab00-8476-11ea-837d-00c6663eaa70.png)

Torres-ssf commented on main.rb:
```diff
@@ -1,17 +1,160 @@
+    arr
+  end
+
+  def my_inject(arg = nil, arg_2 = nil, &prc)
+    array = to_a
+    ind = 0
+    if block_given? == false
+      if arg.is_a?(Symbol)
+        while ind < array.size - 1
+          array[0] = array[0].send(arg, array[ind + 1])
+          ind += 1
+        end
+      elsif arg.is_a?(Integer)
+        while ind < array.size
```
- my_inject. Checkout the [documentation](https://apidock.com/ruby/Enumerable/inject)

- [x] my_inject combines all elements of enum by applying a binary operation, specified by a block.
![Screenshot from 2020-04-22 08-47-15](https://user-images.githubusercontent.com/30977845/79978198-e5395000-8475-11ea-94a0-e0bf4655543b.png)

- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method
![Screenshot from 2020-04-22 08-50-33](https://user-images.githubusercontent.com/30977845/79978473-5d077a80-8476-11ea-98f2-04c351ee2106.png)

expjazz commented:
## Made all the changes required

- The inject method is working as the way required (as seen in the test cases below)
- Made the necessary adjustments to my_none? method

![tse3](https://user-images.githubusercontent.com/61479648/79983495-3fd6aa00-847e-11ea-9331-6a232ab950a4.png)

suretrust commented on .byebug_history:
```diff
@@ -0,0 +1,13 @@
+s
+display arg
+s
+display ind
+s
+display array
+s
+display &prc
+display prc
+display arg
+display ind
+display array
+s
```
Files not related to the project should not be committed and added to .gitignore ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/gitignore.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

- [ ] Delete `.byebug_history` from the repo and add it to `.gitignore`.

suretrust commented on main.rb:
```diff
@@ -1,17 +1,141 @@
+require 'byebug'
```
Once you are done debugging, you should remove the `require 'byebug'`.
